pc2ccs/pc2v9
# I232 submit judges' solutions with tracking

### Description of what the PR does
Allows a judge or administrator to submit the judges' sample submissions and provides a table to compare the expected results with the actual results.  Also provides detailed timings for each test case, as well as executes all test cases, even on failure.  This is necessary for a judge to validate their submissions.

Note: Some of the code in this PR came from the last branch that @lane55 was working on.  I used most of that code, and added more to finish off the PR.  Specifically, all the _list_ related code, as well as the **QuickSubmitter** stuff was from @lane55's branch.  He had the start of a GUI but I pretty much refactored that.

### Issue which the PR addresses
Fixes #232
Fixes #974 

### Environment in which the PR was developed (OS,IDE, Java version, etc.)
OS: Windows 11
java version "1.8.0_321"
Java(TM) SE Runtime Environment (build 1.8.0_321-b07)
Java HotSpot(TM) 64-Bit Server VM (build 25.321-b07, mixed mode)

### Precise steps for _testing_ the PR (i.e., how to demonstrate that it works correctly)
1) Load a contest that has judges' sample solutions - _NAC2024 Dress_ is such a contest.
2) Start the `pc2server`
3) Start up one or more _autojudges_
4) Start up an `pc2admin `client
5) Start the contest
6) Navigate to the **Run->Submit Samples** tab.
7) Open the tab
8) You can choose to **Filter** some submissions by pressing the filter button and selecting what it is you're looking for, or just submit all the runs.  For _NAC 2024 Dress_ there were 68 runs, and it took about 15 mins for a single _autojudge_ to judge them.
9) For this example, I chose to submit just the sample solutions for the first problem.
<img width="797" alt="image" src="https://github.com/pc2ccs/pc2v9/assets/5657363/12e9c7b5-ce2c-4abd-894e-034ed7ff57fc">

10) Press the **Submit** button and then answer **Yes** to the "**_Submit XXX judge sample submissions?_**"
11) You should see the results appear as they submissions are submitted and judged:
<img width="800" alt="image" src="https://github.com/pc2ccs/pc2v9/assets/5657363/1ba5c8f9-f222-4d56-92e7-fdb7c05dcf9f">

12) Results are color-coded.  Green means the PC2 autojudged run matched what the human judge who wrote the submission expected.  Red mean it did not match what was expected.  Yellow means it could not be determined.  Note in the sample below, **Run Id 318** is yellow because the "**Expected**" result is "**rejected**", which is not a valid CLICS judgment.  The **Status** column indicates PC2 judged it as time limit exceeded:
<img width="794" alt="image" src="https://github.com/pc2ccs/pc2v9/assets/5657363/e68d35c1-e9f3-40c2-ae39-24ee43eee818">

13) If you click on the link in the "**Max Time(s)**" column, a window will pop up showing the results of each test case.  For example, if we click on **Run ID 318**'s time of **1.080s**, the following window appears:
<img width="796" alt="image" src="https://github.com/pc2ccs/pc2v9/assets/5657363/91e35e8b-50a0-452a-9e19-68573f0fd34e">

14) The **Result** column shows the autojudge's disposition of the test case.  You can also click on the links in the **Judge's Data** and **Judge's Output** column to display the contents of the specific judges test data.
